### DISCUSSION TITLE
Dask array object size compared to ray.experimental.array

### DISCUSSION THREAD

**Discussion by photoszzt** [dask/dask#7314] [Q&A]
I'm not sure whether it's appropriate to ask here. I have a matrix multiplication program with two implementations, one uses Dask on ray and the other one uses Ray.experimental.array(I modify its DistArray definition with a more flexible block size here https://github.com/photoszzt/ray/blob/master/python/ray/experimental/array/distributed/core.py#L21-L50). The chunk size is the same. When using ray memory to view the object size, I find that dask's object size is always twice as large as the ray.experimental.array. Does anyone know the reason?

```python
from timeit import default_timer as dtimer
import ray
from ray.util.dask import ray_dask_get
import dask
import dask.array as da
import numpy as np


def main():
    _ = ray.init()
    dask.config.set(scheduler=ray_dask_get)
    chunks=(2500, 2500)
    x = da.random.rand((5000, 5000), chunks=chunks)
    y = da.random.rand((5000, 5000), chunks=chunks)
    start = dtimer()
    z = da.matmul(x, y)
    z_val = z.compute()
    end = dtimer()
    print(z_val)
    print(f"time: {end-start} s")


if __name__ == "__main__":
    main()
```

ray.experimental.array version:
```python
from timeit import default_timer as dtimer
import ray
import ray.experimental.array.distributed as rda
import numpy as np


def main():
    chunks=(2500, 2500)
    x = np.random.rand((5000, 5000))
    y = np.random.rand((5000, 5000))
    x = rda.numpy_to_dist.remote(npx, chunks=chunks)
    y = rda.numpy_to_dist.remote(npy, chunks=chunks)
    z = rda.dot.remote(x, y)
    z_val = ray.get(z)
    z_val = z_val.assemble()
    end = dtimer()
    print(z_val)
    print(f"time: {end-start} s")


if __name__ == "__main__":
    main()

**Comment by quasiben** [comment]
@devin-petersohn do you have any thoughts here ?

**Reply by devin-petersohn** [reply]
I don't concretely know the answer, but it could be because of some compression done in the serialization.

**Reply by photoszzt** [reply]
@devin-petersohn What kind of compression in serialization? For Dask array, it overrides __reduce__ for pickleing. 

```python
def __reduce__(self):
        return (Array, (self.dask, self.name, self.chunks, self.dtype))
```
For ray's array, it seems to use the default implementation. Since they both run on Ray, if there's any compression on the ray's store, it should apply to both of them.

**Reply by photoszzt** [reply]
@quasiben @devin-petersohn Let me be more concrete: In the example code, a 2500*2500 float64 matrix would consume 2500*2500*8/1024/1024 ~= 47 MB. In ray.experimental.array, I'm seeing the object size similar to that. But for dask on ray, the object size is ~= 95 MB which is 2x the correct size. It seems like the object contains two matrix or the same matrix get stored twice.

**Reply by photoszzt** [reply]
Hmm, I run the code one more time with ray master. Now it’s not 2 times. It’s much closer. dask uses 810 MB and ray.experimental.array uses 762 MB. Not sure what changed.
